launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Launcher for Spark applications.
+ * <p/>
+ * Use this class to start Spark applications programmatically. The class uses a builder pattern
+ * to allow clients to configure the Spark application and launch it as a child process.
+ * <p/>
+ * Note that launching Spark applications using this class will not automatically load environment
+ * variables from the "spark-env.sh" or "spark-env.cmd" scripts in the configuration directory.
+ */
+public class SparkLauncher extends AbstractLauncher<SparkLauncher> {
+
+  boolean verbose;
+  String appName;
+  String master;
+  String deployMode;
+  String mainClass;
+  String appResource;
+  final List<String> sparkArgs;
+  final List<String> appArgs;
+  final List<String> jars;
+  final List<String> files;
+  final List<String> pyFiles;
+
```

Comment:
since this is the public interface it might be nice to do full javadoc on all the public function in here describing parameters and return values, etc..